src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1845,20 +1880,84 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
+                call_name = at->m_attr;
+                ASR::symbol_t *t = m->m_symtab->resolve_symbol(call_name);
+                if (!t) {
+                    throw SemanticError("The symbol '" + call_name + "' not found in the module '" + \
+                        mod_name + "'", x.base.base.loc);
+                }
+                call_name = "__" + mod_name + "_" + call_name;
+                if (ASR::is_a<ASR::Subroutine_t>(*t)) {
+                    ASR::Subroutine_t *msub = ASR::down_cast<ASR::Subroutine_t>(t);
+                    Str name;
+                    name.from_str(al, call_name);
+                    ASR::asr_t *sub = ASR::make_ExternalSymbol_t(
+                        al, msub->base.base.loc,
+                        /* a_symtab */ current_scope,
+                        /* a_name */ name.c_str(al),
+                        (ASR::symbol_t*)msub,
+                        m->m_name, nullptr, 0, msub->m_name,
+                        ASR::accessType::Public
+                        );
+                    ASR::symbol_t *st = ASR::down_cast<ASR::symbol_t>(sub);
+                    current_scope->scope[call_name] = st;
+                    tmp = ASR::make_SubroutineCall_t(al, x.base.base.loc, st,
+                        nullptr, args.p, args.size(), nullptr);
+                    return;
+                } else if (ASR::is_a<ASR::Function_t>(*t)) {
+                    ASR::Function_t *mfn = ASR::down_cast<ASR::Function_t>(t);
+                    Str name;
+                    name.from_str(al, call_name);
+                    char *cname = name.c_str(al);
+                    ASR::asr_t *fn = ASR::make_ExternalSymbol_t(
+                        al, mfn->base.base.loc,
+                        /* a_symtab */ current_scope,
+                        /* a_name */ cname,
+                        (ASR::symbol_t*)mfn,
+                        m->m_name, nullptr, 0, mfn->m_name,
+                        ASR::accessType::Public
+                        );
+                    ASR::symbol_t *st = ASR::down_cast<ASR::symbol_t>(fn);
+                    current_scope->scope[call_name] = st;
+                    ASR::ttype_t *a_type = ASRUtils::expr_type(mfn->m_return_var);
+                    tmp = ASR::make_FunctionCall_t(al, x.base.base.loc, st,
+                        nullptr, args.p, args.size(), nullptr, 0, a_type,
+                        nullptr, nullptr);
+                    return;
+                } else {
+                    throw SemanticError("Only Subroutines and Functions are currently supported in 'import'",
+                        x.base.base.loc);
+                }
+
+            } else {
+                throw SemanticError("Only Name type supported in Call",
+                    x.base.base.loc);
+            }
+        } else {
+            throw SemanticError("Only Name or Attribute type supported in Call",
+                x.base.base.loc);
+        }
```

Comment:
This also looks great, that's how I would do it. I think what remains are bugs to identify and fix.

Reply:
Yep. The main bug is now why `current_scope` is unable to find the module name in `visit_Call`, when we have stored it during `visit_Import`

Reply:
I think that's because `current_scope` is the local scope in the subroutine, and that scope does not have the module. The module lives at the top level.

Reply:
We should figure out how to handle nested modules, so that you can actually import a math module into a subroutine's scope (if you do `import math` inside a subroutine), in Fortran modules are global, and in Python I am currently not sure, they might also be global in some sense, but they can definitely be nested inside other modules and subroutines. But for now, let's implement the global approach. For that, just do something like:
```
while (symtab->parent != nullptr) symtab = symtab->parent;
```
That gives you the global scope symbol table. And then from there, just find the `math` module, it should be in this table.

Reply:
Okay, trying that!